tests/test_qutrit_device.py

```diff
@@ -0,0 +1,1022 @@
+
+        assert np.array_equal(out, expected_samples)
+
+
+class TestGenerateSamples:
+    """Test the generate_samples method"""
+
+    def test_auxiliary_methods_called_correctly(self, mock_qutrit_device, monkeypatch):
+        """Tests that the generate_samples method calls on its auxiliary methods correctly"""
+
+        dev = mock_qutrit_device()
+        number_of_states = 3**dev.num_wires
+
+        with monkeypatch.context() as m:
+            # Mock the auxiliary methods such that they return the expected values
+            m.setattr(QutritDevice, "sample_basis_states", lambda self, wires, b: wires)
+            m.setattr(QutritDevice, "states_to_ternary", lambda a, b: (a, b))
+            m.setattr(QutritDevice, "analytic_probability", lambda *args: None)
+            m.setattr(QutritDevice, "shots", 1000)
+            dev._samples = dev.generate_samples()
+
+        assert dev._samples == (number_of_states, dev.num_wires)
+
+
+class TestSampleBasisStates:
+    """Test the sample_basis_states method"""
+
+    def test_sampling_with_correct_arguments(self, mock_qutrit_device, monkeypatch):
+        """Tests that the sample_basis_states method samples with the correct arguments"""
+
+        shots = 1000
+
+        number_of_states = 9
+        dev = mock_qutrit_device()
+        dev.shots = shots
+        state_probs = [0.1] * 9
+        state_probs[0] = 0.2
+
+        with monkeypatch.context() as m:
+            # Mock the numpy.random.choice method such that it returns the expected values
+            m.setattr("numpy.random.choice", lambda x, y, p: (x, y, p))
+            res = dev.sample_basis_states(number_of_states, state_probs)
+
+        assert np.array_equal(res[0], np.array([0, 1, 2, 3, 4, 5, 6, 7, 8]))
+        assert res[1] == shots
+        assert res[2] == state_probs
+
+    def test_raises_deprecation_warning(self, mock_qutrit_device, monkeypatch):
+        """Test that sampling basis states on a device with shots=None produces a warning."""
+
+        dev = mock_qutrit_device()
+        number_of_states = 9
+        dev.shots = None
+        state_probs = [0.1] * 9
+        state_probs[0] = 0.2
+
+        with pytest.raises(
+            qml.QuantumFunctionError,
+            match="The number of shots has to be explicitly set on the device",
+        ):
+            dev.sample_basis_states(number_of_states, state_probs)
+
+
+class TestStatesToTernary:
+    """Test the states_to_ternary method"""
+
+    def test_correct_conversion_two_states(self, mock_qutrit_device):
+        """Tests that the sample_basis_states method converts samples to binary correctly"""
+        wires = 4
+        shots = 10
+
+        number_of_states = 3**wires
+        basis_states = np.arange(number_of_states)
+        samples = np.random.choice(basis_states, shots)
+
+        dev = mock_qutrit_device()
+        res = dev.states_to_ternary(samples, wires)
+
+        expected = []
+
+        for s in samples:
```

Comment:
This looks like essentially the same code that is being used within the `states_to_ternary` method; so if there is an error in the methodology, it wouldn't actually be caught. Recommend to either hardcode some examples in here, and/or to use a different built-in method in order to generate the expected values.

Reply:
That is a good point. I'll change the test to use hardcoded examples.